### PR TITLE
HTTP/2 Multipart message handler

### PR DESCRIPTION
**Pull request by s-maheshbabu**
AVS returns HTTP/2 Multipart messages to it clients. See https://developer.amazon.com/docs/alexa-voice-service/structure-http2-request.html for more details on the exact structure.

Added a parser that takes the multi-part reponse and extracts the 'caption' which represents the actual content that should be displayed to the user as Alexa's response.

The rest of the message contains audio bytes of the same response, SSML tagsetc., all of which gets dropped.

We are using https://github.com/miguelmota/http-message-parser/ to do the parsing.

Unit tests added.

Note: http-message-parser has a bug in the latest version and so package.json is configured to not use the latest version of this package but use 0.0.22.
A bug has been filed against the developer at miguelmota/http-message-parser#4

### CODE REVIEW

File: src/SpeakDirectiveParser.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    );
+  }
+
+  const avsDirectiveBuffer = parsedResponse.multipart[0].body;
+  let avsDirective;
+  try {
+    avsDirective = JSON.parse(avsDirectiveBuffer);
+  } catch (e) {
+    console.error(
+      "JSON parse error: " + util.inspect(e, { showHidden: true, depth: null })
```
Comment (gootla-manasa) [inline src/SpeakDirectiveParser.js]: Why are you logging the error as well as throwing it? Have this information in the IllegalArgumentError which can be used by the callers of this function and remove this line.

File: src/SpeakDirectiveParser.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ */
+export function extractAlexaTextResponse(alexaRawResponse) {
+  if (!alexaRawResponse) {
+    throw new IllegalArgumentError(
+      "The response to be parsed cannot be empty. Input: " + alexaRawResponse
+    );
+  }
+
+  const parsedResponse = httpMessageParser(alexaRawResponse);
+  if (
```
Comment (gootla-manasa) [inline src/SpeakDirectiveParser.js]: Why dont you extract extractive avsDirective from parsed Response into a private method? That way you can test other failing cases (eg. parseArgument undefined or empty) which you failed to capture as of today.

File: src/SpeakDirectiveParser.test.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
+import { extractAlexaTextResponse } from "./SpeakDirectiveParser";
+import IllegalArgumentError from "./errors/IllegalArgumentError";
+
+import testData from "./test-data/multipart-response-test-data";
+
+it("throws an error if an empty string is passed as input", () => {
```
Comment (gootla-manasa) [inline src/SpeakDirectiveParser.test.js]: Add tests to handle empty/undefined results from *httpMessageParser*. Also test cases where the key paths for parsed message are missing from the response.

File: src/test-data/multipart-response-test-data.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
+// Invalid case where multi-part message has just one part. This should never happen in real world.
+{
+  const rawData = String.raw`--------abcde123
+Content-Type: application/json; charset=UTF-8
```
Comment (gootla-manasa) [inline src/test-data/multipart-response-test-data.js]: Please maintain indentation here. The dataset of constants is confusing.

File: src/test-data/multipart-response-test-data.js [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+second-part-of-multi-part-message
+--------abcde123--`;
+
+  exports.directive_not_valid_json = {
+    rawData: rawData
+  };
+}
+
+// Invalid case where the AVS directive is well formatted but doesn't contain the 'directive' key.
```
Comment (gootla-manasa) [inline src/test-data/multipart-response-test-data.js]: What do you mean by well formatted directive?

File: src/SpeakDirectiveParser.js (hunk 1 above)
Comment (s-maheshbabu) [reply src/SpeakDirectiveParser.js]: Done.

File: src/test-data/multipart-response-test-data.js (hunk 4 above)
Comment (s-maheshbabu) [reply src/test-data/multipart-response-test-data.js]: Unfortunately, it can't be done. The multi-part format won't tolerate unnecessary spaces.

File: src/test-data/multipart-response-test-data.js (hunk 5 above)
Comment (s-maheshbabu) [reply src/test-data/multipart-response-test-data.js]: As in, json formatted. Added the word json.

File: src/SpeakDirectiveParser.js (hunk 2 above)
Comment (s-maheshbabu) [reply src/SpeakDirectiveParser.js]: That feels like completely giving up and is a workaround anyways. At some point, I want to figure out how to mock httpParser for real. 

How about a TODO that says, we need to add tests by mocking httpMessageParser? It should be safe because httpMessageParser never returns empty response or empty response.multipart.body and it is a library that we can upgrade whenever we want to.

File: src/SpeakDirectiveParser.js [hunk 6]
```diff
@@ -37,12 +37,10 @@ export function extractAlexaTextResponse(alexaRawResponse) {
...
     avsDirective = JSON.parse(avsDirectiveBuffer);
   } catch (e) {
-    console.error(
-      "JSON parse error: " + util.inspect(e, { showHidden: true, depth: null })
-    );
     throw new IllegalArgumentError(
-      "Given directive couldn't be parsed to a JSON object. Input: " +
-        avsDirectiveBuffer.toString()
+      `Given directive couldn't be parsed to a JSON object. Input: " ${avsDirectiveBuffer.toString()}
+      StackTrace:
```
Comment (gootla-manasa) [inline src/SpeakDirectiveParser.js]: Does the error actually give out a stack trace? Did you verify it? If it is not a stack trace, then the error message needs to be updated.

File: src/SpeakDirectiveParser.js (hunk 2 above)
Comment (gootla-manasa) [reply src/SpeakDirectiveParser.js]: Okay.

File: src/SpeakDirectiveParser.js (hunk 6 above)
Comment (s-maheshbabu) [reply src/SpeakDirectiveParser.js]: Yep. Verified by putting this line before the throw statement and it does print the stack trace. 

```javascript
console.log(`Given directive couldn't be parsed to a JSON object. Input: " ${avsDirectiveBuffer.toString()}
      StackTrace:
      ${util.inspect(e, { showHidden: true, depth: null })}`);
```

File: src/SpeakDirectiveParser.js (hunk 2 above)
Comment (s-maheshbabu) [reply src/SpeakDirectiveParser.js]: Done.

File: src/SpeakDirectiveParser.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+  const avsDirectiveBuffer = parsedResponse.multipart[0].body;
+  let avsDirective;
+  try {
+    avsDirective = JSON.parse(avsDirectiveBuffer);
+  } catch (error) {
+    throw new IllegalArgumentError(
+      `Given directive couldn't be parsed to a JSON object. Input: " ${avsDirectiveBuffer.toString()}
+      StackTrace:
+      ${util.inspect(error, { showHidden: true, depth: null })}`
```
Comment (gootla-manasa) [inline src/SpeakDirectiveParser.js]: Could you verify here that you are able to get the stack trace as intended? A manual check is good enough for now.

Comment (s-maheshbabu) [reply src/SpeakDirectiveParser.js]: Didn't I already address the same comment?

https://github.com/s-maheshbabu/silent-alexa/pull/9/commits/1f93168982fc4a781e7ad3a9035df91c53146be4
